File: Data Structures & Algorithms/palindromic-substrings/submission-0.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int n = s.size(), res = 0;
        vector<vector<int>> memo(n, vector<int>(n, -1));
        function<bool(int, int)> isPalindrome = [&](int i, int j) -> bool {
            if (i >= j) return true;
            if (s[i] != s[j]) return false;
            if (memo[i][j] != -1) return memo[i][j];
            return memo[i][j] = isPalindrome(i + 1, j - 1);
        };
        for (int i = 0; i < n; i++) {
            for (int j = n - 1; j >= 0; j--) {
                if (i <= j) {
                    if (isPalindrome(i, j)) {
                        res++;
                    }
                }
            }
        }
        return res;
    }
};

File: Data Structures & Algorithms/is-palindrome/submission-0.py
class Solution:
    def isAlphaNum(self, c):
        return ord('A') <= ord(c) <= ord('Z') or ord('a') <= ord(c) <= ord('z') or ord('0') <= ord(c) <= ord('9')

    def isPalindrome(self, s: str) -> bool:
        i, j = 0, len(s) - 1
        lowerS = s.lower()
        while i < j:
            if not self.isAlphaNum(s[i]):
                i += 1
                continue
            if not self.isAlphaNum(s[j]):
                j -= 1
                continue
            if lowerS[i] != lowerS[j]:
                return False
            i += 1
            j -= 1
        return True